File: src/15_Spanning_Trees/kruskal.cpp
#include <bits/stdc++.h>
using namespace std;


const int m = 8;
vector<tuple<int, int, int> > edge_list = {
  make_tuple(1, 2, 3),
  make_tuple(1, 5, 5),
  make_tuple(2, 5, 6),
  make_tuple(2, 3, 5),
  make_tuple(5, 6, 2),
  make_tuple(3, 6, 3),
  make_tuple(3, 4, 9),
  make_tuple(6, 4, 7)
};


// apparently link is reserved
int arc[m];
int set_size[m];


// O(logn)
int representative(int x)
{
  while (x != arc[x]) x = arc[x];
  return x;
}


// O(logn)
bool same(int a, int b) {return representative(a) == representative(b);}


// O(logn)
// Makes sure that the length of each chain is O(logn), by connecting the
// smaller set to the larger set.
void unite(int a, int b)
{
  a = representative(a);
  b = representative(b);
  if (set_size[a] < set_size[b]) swap(a, b);
  set_size[a] += set_size[b];
  arc[b] = a;
}


struct shorter_edge
{
  bool operator()(tuple<int, int, int> e1, tuple<int, int, int> e2)
  {
    return get<2>(e1) < get<2>(e2);
  }
};


// O(mlog(n)) if edge_list is sorted else O(mlog(m))
void kruskal()
{
  sort(edge_list.begin(), edge_list.end(), shorter_edge());  // O(mlog(m))
  iota(arc, arc + m, 0);
  fill(set_size, set_size + m, 1);
  int cost = 0;

  cout << "MST:\n";
  for (auto&& [a, b, w] : edge_list) {  // O(mlog(n))
    if (!same(a, b)) {
      cost += w;
      unite(a, b);
      // this is an MST edge
      cout << a << " - " << b << '\n';
    }
  }
  cout << "total cost: " << cost << '\n';
}


int main() {
  kruskal();
}

File: src/15_Spanning_Trees/kruskal.py

from operator import itemgetter

edge_list = [
  [1, 2, 3],
  [1, 5, 5],
  [2, 5, 6],
  [2, 3, 5],
  [5, 6, 2],
  [3, 6, 3],
  [3, 4, 9],
  [6, 4, 7]
]
m = len(edge_list)

# Union-Find or Disjoint Set Union (DSU)
# m should be >= n - 1, so it can be used as length, if n is unknown
link = list(range(m + 1))
size = [1] * (m + 1)


# log(n)
def find(x):
  """Finds the representative of x."""
  while x != link[x]:
    x = link[x]
  return x


# log(n)
def same(a, b):
  """Checks if a, b belong to the same set."""
  return find(a) == find(b)


# log(n)
def unite(a, b):
  """Attach the smaller set to the larger (union by size).

  The depth of most nodes (larger set) stays the same.

  Why log2(n)?
  ------------
  Initially, each element starts as a single-node tree with rank 0. When two
  trees of the same rank r get merged, the resulting rank is r + 1 and the size
  is roughly doubled. Namely, the size of a tree with rank r is roughly 2^r.
  Since the nodes in the set can be at most n, n = 2^r => r = log2(n), which is
  the number of hops to get to the represenative. In practice, all operations
  of Union-Find are much faster, closer to O(1).
  """
  a = find(a)
  b = find(b)
  if size[a] < size[b]:
    a, b = b, a
  size[a] += size[b]
  link[b] = a


# O(mlog(n)) if edge_list is sorted else O(mlog(m))
def kruskal():
  edge_list.sort(key=itemgetter(2))
  cost = 0
  print("MST:")
  for a, b, c in edge_list:
    if not same(a, b):
      # then, this is an MST edge
      cost += c
      unite(a, b)
      print(a, '-', b)
  print("total cost:", cost)


if __name__ == '__main__':
  kruskal()
